### PR TITLE
Display mesh faces from MDAL [3D]

### PR DESCRIPTION
**Pull request by PeterPetrik**
I do not see reason why we cannot show 2DM and various other terrain profile from MDAL in 3D. Really, why not?

screenshot 2019-01-09 at 16 28 18

### CODE REVIEW

**Comment by nirvn** [comment]
Now that's just cool :)

**Comment by nirvn** [comment]
On a totally unrelated front: wow those color button widgets aren't rendering well at all under Qt's macintosh style.

**Comment by saberraz** [comment]
That is a very nice feature. I have tried it with a mesh layer containing multiple quantities. But it only uses the first one for 3d scene.

File: src/3d/symbols/qgspolygon3dsymbol_p.cpp [hunk 1]
```diff
@@ -224,4 +254,53 @@ Qt3DRender::QGeometryRenderer *QgsPolygon3DSymbolEntityNode::renderer( const Qgs
...
+      QgsPolygon *polygon = QgsMeshUtils::toPolygon( triangle, vertices );
+      Qgs3DUtils::clampAltitudes( polygon, symbol.altitudeClamping(), symbol.altitudeBinding(), height, map );
+      polygons.append( polygon );
+      fids.append( i );
+      if ( hasDDExtrusion )
+        extrusionHeightPerPolygon.append( extrusionHeight );
+    }
+  }
+
+  // TODO: well, those polygons are already triangles
```
Comment (nyalldawson) [inline src/3d/symbols/qgspolygon3dsymbol_p.cpp]: What's this Todo actually mean?

File: src/core/mesh/qgstriangularmesh.h [hunk 2]
```diff
@@ -145,6 +145,9 @@ namespace QgsMeshUtils
   //! Returns face as polygon geometry
   QgsGeometry toGeometry( const QgsMeshFace &face, const QVector<QgsMeshVertex> &vertices );
 
+  //! Returns face as polygon geometry, caller is responsible for delete
+  CORE_EXPORT QgsPolygon *toPolygon( const QgsMeshFace &face, const QVector<QgsMeshVertex> &vertices );
```
Comment (nyalldawson) [inline src/core/mesh/qgstriangularmesh.h]: Either SIP_FACTORY, or return a unique_ptr

**Comment by PeterPetrik** [comment]
> That is a very nice feature. I have tried it with a mesh layer containing multiple quantities. But it only uses the first one for 3d scene.

to use other dataset groups and for example see the flood progress in 3D would need a lot of more work, e.g. implementation of time slider , group selection , how to visualize river/water (polygons or volumes), what to do with vectors (3d arrows?)

File: src/3d/qgsmeshlayer3drenderer.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMESHLAYER3DRENDERER_H
+#define QGSMESHLAYER3DRENDERER_H
+
+#define SIP_NO_FILE
```
Comment (wonder-sk) [inline src/3d/qgsmeshlayer3drenderer.h]: Vector layer 3D renderer now has (experimental) Python support, so best to enable Python API for mesh 3d renderer too

File: src/3d/symbols/qgspolygon3dsymbol_p.cpp [hunk 4]
```diff
@@ -224,4 +254,55 @@ Qt3DRender::QGeometryRenderer *QgsPolygon3DSymbolEntityNode::renderer( const Qgs
...
+    for ( int i = 0; i < triangles.size(); ++i )
+    {
+      const QgsMeshFace &triangle = triangles.at( i );
+      Q_ASSERT( triangle.size() == 3 );
+      std::unique_ptr< QgsPolygon > polygon = QgsMeshUtils::toPolygon( triangle, vertices );
+      Qgs3DUtils::clampAltitudes( polygon.get(), symbol.altitudeClamping(), symbol.altitudeBinding(), height, map );
+      polygons.append( polygon.release() );
+      fids.append( i );
+      if ( hasDDExtrusion )
+        extrusionHeightPerPolygon.append( extrusionHeight );
```
Comment (wonder-sk) [inline src/3d/symbols/qgspolygon3dsymbol_p.cpp]: I am not sure if extrusion makes sense for mesh - I would prefer not to use it and not even show it in the GUI

File: src/3d/symbols/qgspolygon3dsymbol_p.cpp [hunk 5]
```diff
@@ -224,4 +254,55 @@ Qt3DRender::QGeometryRenderer *QgsPolygon3DSymbolEntityNode::renderer( const Qgs
...
+    }
+  }
+
+  // Polygons from mesh are already triangles, but
+  // call QgsTessellatedPolygonGeometry to
+  // use symbol settings for back faces, normals, etc
+  mGeometry = new QgsTessellatedPolygonGeometry;
+  mGeometry->setInvertNormals( symbol.invertNormals() );
```
Comment (wonder-sk) [inline src/3d/symbols/qgspolygon3dsymbol_p.cpp]: inversion of normals also does not make sense here in my opinion - that's used just for polygon data when they have the reversed vertex ordering of individual polygons.

**Review by wonder-sk — COMMENTED** [review]
Nice feature!

The usage of 3D polygon symbols will soon need to be replaced by a custom 3D mesh symbol because we will need a special material (to draw per-vertex colors) and thus also special geometry class too. But for now it is a good shortcut to get meshes displayed in 3D!

**Pull request merged**